### PR TITLE
Add a single URLStreamHandler for all resources

### PR DESCRIPTION
**Pull request by johanvos**
This addresses #1030

### CODE REVIEW

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/Resources.java [hunk 1]
```diff
@@ -134,4 +136,31 @@ protected URLConnection openConnection(URL u) throws IOException {
...
+            protected URLConnection openConnection(URL url) throws IOException {
+                return new URLConnection(url) {
+                    @Override
+                    public void connect() throws IOException {
+                    }
+
+                    @Override
+                    public InputStream getInputStream() throws IOException {
+                        ResourcesSupport support = ImageSingletons.lookup(ResourcesSupport.class);
+                        String resName = url.toString().substring(9);
```
Comment (cstancu) [inline substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/Resources.java]: Please add a comment explaining the value `9`.

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/LocalizationSupport.java [hunk 2]
```diff
@@ -95,12 +96,13 @@ void includeResourceBundles() {
...
      */
-    public ResourceBundle getCached(String baseName, Locale locale) {
+    public ResourceBundle getCached(String baseName, Locale locale) throws MissingResourceException {
         ResourceBundle result = cache.get(baseName);
         if (result == null) {
             String errorMessage = "Resource bundle not found " + baseName + ". " +
                             "Register the resource bundle using the option " + includeResourceBundlesOption + baseName + ".";
-            throw VMError.unsupportedFeature(errorMessage);
+            System.err.println(errorMessage);
+            throw new MissingResourceException(errorMessage, this.getClass().getName(), baseName);//VMError.unsupportedFeature(errorMessage);
```
Comment (cstancu) [inline substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/LocalizationSupport.java]: Please remove the comment at the end of the line.

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/LocalizationSupport.java [hunk 3]
```diff
@@ -95,12 +96,13 @@ void includeResourceBundles() {
...
      * @param locale this parameter is not currently used.
      */
-    public ResourceBundle getCached(String baseName, Locale locale) {
+    public ResourceBundle getCached(String baseName, Locale locale) throws MissingResourceException {
         ResourceBundle result = cache.get(baseName);
         if (result == null) {
             String errorMessage = "Resource bundle not found " + baseName + ". " +
                             "Register the resource bundle using the option " + includeResourceBundlesOption + baseName + ".";
-            throw VMError.unsupportedFeature(errorMessage);
+            System.err.println(errorMessage);
```
Comment (cstancu) [inline substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/LocalizationSupport.java]: Is the `System.err.println` needed given that you throw a `MissingResourceException` below?

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/Resources.java [hunk 4]
```diff
@@ -107,6 +108,7 @@ public static void registerResource(String name, InputStream is) {
         return ImageSingletons.lookup(ResourcesSupport.class).resources.get(name);
     }
 
+
```
Comment (cstancu) [inline substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/Resources.java]: Please remove the extra line.

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/Resources.java [hunk 5]
```diff
@@ -56,6 +56,7 @@
 
     static class ResourcesSupport {
         final Map<String, List<byte[]>> resources = new HashMap<>();
+        final URLStreamHandler resourcesURLStreamHandler = createResourcesURLStreamHandler();
```
Comment (cstancu) [inline substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/Resources.java]: I don't think that we need the `ResourceSupport.resourcesURLStreamHandler` field. The `Resources.createResourcesURLStreamHandler()` method could be moved to `JavaNetSubstitutions` instead.

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/JavaNetSubstitutions.java [hunk 6]
```diff
@@ -131,6 +133,10 @@ static URLStreamHandler get(String protocol) {
     @Platforms(Platform.HOSTED_ONLY.class)
     static boolean addURLStreamHandler(String protocol) {
         try {
+            if (RESOURCE_PROTOCOL.equals(protocol)) {
+                URLProtocolsSupport.put(RESOURCE_PROTOCOL,ImageSingletons.lookup(Resources.ResourcesSupport.class).resourcesURLStreamHandler);
+                return true;
+            }
```
Comment (cstancu) [inline substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/JavaNetSubstitutions.java]: I think this block can be moved outside the `try`.

**Comment by cstancu** [comment]
@johanvos this PR contains commits from two different email addresses: `johan@lodgon.com` and `johan.vos@gluonhq.com`. Can you please squash them and use the email that you signed the OCA for, i.e., `johan@lodgon.com`? Thanks!

**Comment by cstancu** [comment]
@johanvos I cannot merge these changes before the OCA check issue is fixed.

**Comment by johanvos** [comment]
I have no idea why the OCA check fails here, as it passed on all my other PR's?

**Comment by dougxc** [comment]
I think it's because you made at least one of the commits in this PR with a different email address. I've added it to the OCA bot database so it should succeed the check in a few minutes.

**Comment by cstancu** [comment]
One of the commits seems to be using a different email address than the last one you signed the OCA for, namely `johan.vos@gluonhq.com`.

**Comment by johanvos** [comment]
Right, now I see it indeed. I can create a new PR if you want, as I made it uglier when trying to change the email address with rebase.

**Pull request closed without merging**